tokio/src/sync/semaphore_ll.rs

```diff
@@ -501,86 +596,149 @@ impl Permit {
+        use PermitState::*;
+
         match self.state {
-            PermitState::Idle => {}
-            PermitState::Waiting => {
+            Waiting(requested) => {
+                // There must be a waiter
                 let waiter = self.waiter.as_ref().unwrap();
 
-                if waiter.acquire(cx)? {
-                    self.state = PermitState::Acquired;
+                match requested.cmp(&num_permits) {
+                    Less => {
+                        let delta = num_permits - requested;
+
+                        // Request additional permits. If the waiter has been
+                        // dequeued, it must be re-queued.
+                        if !waiter.try_inc_permits_to_acquire(delta as usize) {
+                            let waiter = NonNull::from(&**waiter);
+
+                            // Ignore the result. The check for
+                            // `permits_to_acquire()` will converge the state as
+                            // needed
+                            let _ = semaphore.poll_acquire2(delta, || Some(waiter))?;
+                        }
+
+                        self.state = Waiting(num_permits);
+                    }
+                    Greater => {
+                        let delta = requested - num_permits;
+                        let to_release = waiter.try_dec_permits_to_acquire(delta as usize);
+
+                        semaphore.add_permits(to_release);
+                        self.state = Waiting(num_permits);
+                    }
+                    Equal => {}
+                }
+
+                if waiter.permits_to_acquire()? == 0 {
+                    self.state = Acquired(requested);
+                    return Ready(Ok(()));
+                }
+
+                waiter.waker.register_by_ref(cx.waker());
+
+                if waiter.permits_to_acquire()? == 0 {
+                    self.state = Acquired(requested);
                     return Ready(Ok(()));
-                } else {
-                    return Pending;
                 }
-            }
-            PermitState::Acquired => {
-                return Ready(Ok(()));
-            }
-        }
 
-        match semaphore.poll_permit(Some((cx, self)))? {
-            Ready(()) => {
-                self.state = PermitState::Acquired;
-                Ready(Ok(()))
-            }
-            Pending => {
-                self.state = PermitState::Waiting;
                 Pending
             }
+            Acquired(acquired) => {
+                if acquired >= num_permits {
+                    Ready(Ok(()))
+                } else {
+                    match semaphore.poll_acquire(cx, num_permits - acquired, self)? {
+                        Ready(()) => {
+                            self.state = Acquired(num_permits);
+                            Ready(Ok(()))
+                        }
+                        Pending => {
+                            self.state = Waiting(num_permits);
```

Comment:
Do we need to store somewhere how many permits have already been acquired? Not sure if there is still a `Waiter` with the previous information, since in the `Acquired` state it might already have been removed?

Reply:
The number of permits acquired can be computed by taking the number of requested permits and subtract the number of permits remaining to be acquired (stored in the `Waiter` atomic state).